### PR TITLE
Support choice observers in 1.dev

### PR DESCRIPTION
**Pull request by nickchapman-da**
Done:
- update hash for ghc patch version
- update stack-snapshot hashes for ghc-lib(-parser)
- update stackage_snapshot.json, following `bazel run @stackage-unpinned//:pin`
- expose Optional constructors (Some/None) for desugared code to use
- adapt LFConversion to expect a 4-tuple for a desugared choice def/sig
- adapt LFConversion to convert optional choice-observers function
- add first example which uses choice-observers
- add `jq` tests the choice-observers example
- add test for choice-observer behaviour: `DiscloseViaChoiceObserver`
- update stackage_snapshot.json, following `bazel run @stackage-unpinned//:pin` (on windows) (Thanks @garyverhaegen-da)
- updated documentation for choice observers

changelog_begin
Support choice observers in 1.dev
issue link: Add choice observers #7709
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala [hunk 1]
```diff
@@ -359,7 +359,8 @@ private[validation] object Typing {
             checkExpr(controllers, TParties)
           }
           choiceObservers.foreach {
-            checkExpr(_, TParties) // FIXME #7709, be conditional on: supportsContractObservers
+            introExprVar(param, paramType)
```
Comment (hurryabit) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:362]: Good catch!

**Review by hurryabit — COMMENTED** [review]
Looks good so far.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala (hunk 1 above)
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:362]: Thanks to Moritz for pointing me in the right direction

File: compiler/damlc/tests/daml-test-files/ChoiceObservers.daml [hunk 2]
```diff
@@ -0,0 +1,40 @@
+-- Copyright (c) 2020, Digital Asset (Switzerland) GmbH and/or its affiliates.
+-- All rights reserved.
+
+module ChoiceObservers where
+
+-- @SINCE-LF 1.dev
+
+-- @QUERY-LF [ .modules[] | .templates[] | select(lf::get_template_name($pkg) == ["TheTemplate"]) | .choices[] | select(lf::get_name($pkg) == "C1") | .observers ] == [ null ]
+
+-- @QUERY-LF [ .modules[] | .templates[] | select(lf::get_template_name($pkg) == ["TheTemplate"]) | .choices[] | select(lf::get_name($pkg) == "C2") | .observers ] != [ null ]
```
Comment (hurryabit) [inline compiler/damlc/tests/daml-test-files/ChoiceObservers.daml:10]: It's probably safer to make a positive statement about the observers field here, something like "it's an object with a field name `foo`". Things can be different for so many reasons.

**Review by hurryabit — COMMENTED** [review]
The `jq` tests look generally good. One small idea how to make them even better.

File: compiler/damlc/tests/daml-test-files/ChoiceObservers.daml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      do return ()
+
+    -- New syntax: choice with observer and controller.
+    choice C2 : () with ys : [Party]
+      observer ys
+      controller ys
+      do return ()
+
+    choice C3 : () with ys : [Party]
+      observer ([]:[Party]) -- explicit empty choice observers; should be disnguishable from no observer clause
```
Comment (hurryabit) [inline compiler/damlc/tests/daml-test-files/ChoiceObservers.daml:33]:
```suggestion
      observer ([]:[Party]) -- explicit empty choice observers; should be distinguishable from no observer clause
```

File: compiler/damlc/tests/daml-test-files/ChoiceObservers.daml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      controller ys
+      do return ()
+
+    choice C4 : () with ys : [Party]
+      observer [this.s, s, another] ++ arg.ys ++ ys -- check everything is correctly in scope
```
Comment (hurryabit) [inline compiler/damlc/tests/daml-test-files/ChoiceObservers.daml:38]: Excellent idea to check for variables bound in template-level `let`s as well.

File: compiler/damlc/tests/daml-test-files/DiscloseViaChoiceObserver.daml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      observer receiver
+      controller owner
+      do pure ()
+    nonconsuming choice Reveal: Text with
+        receiver: Party
+      controller receiver
+      do pure text
+
+test : Scenario ()
+test = scenario do
```
Comment (hurryabit) [inline compiler/damlc/tests/daml-test-files/DiscloseViaChoiceObserver.daml:26]: This checks only that we divulge the contract which contains the choice with an observer. We should also add a test for checking that we also divulge everything happening in the choice body. Does that make sense?

File: compiler/damlc/tests/daml-test-files/ChoiceObservers.daml (hunk 4 above)
Comment (nickchapman-da) [reply compiler/damlc/tests/daml-test-files/ChoiceObservers.daml:38]: It was your idea! :)

File: compiler/damlc/tests/daml-test-files/DiscloseViaChoiceObserver.daml (hunk 5 above)
Comment (nickchapman-da) [reply compiler/damlc/tests/daml-test-files/DiscloseViaChoiceObserver.daml:26]: You mean, for example, if the choice body fetches some other contracts, these also need to be divulged?

Comment (nickchapman-da) [reply compiler/damlc/tests/daml-test-files/DiscloseViaChoiceObserver.daml:26]: I added another test.

File: compiler/damlc/tests/daml-test-files/MoreChoiceObserverDivulgence.daml [hunk 6]
```diff
@@ -0,0 +1,48 @@
+-- Copyright (c) 2020, Digital Asset (Switzerland) GmbH and/or its affiliates.
+-- All rights reserved.
+
+module MoreChoiceObserverDivulgence where
```
Comment (hurryabit) [inline compiler/damlc/tests/daml-test-files/MoreChoiceObserverDivulgence.daml:4]: Nice test. Thank you.

File: docs/source/concepts/glossary.rst [hunk 7]
```diff
@@ -110,6 +110,11 @@ A **controller** is a `party <#party>`__ that is able to `exercise <#exercise>`_
 
 Controllers must be at least an `observer`_, otherwise they can't see the contract to exercise it on. But they don't have to be a `signatory`_. this enables the :doc:`propose-accept pattern </daml/patterns/initaccept>`.
 
+Choice Observer
+---------------
+
+A **choice observer** is a `party <#party>`__ on a `choice <#choice>`__. Being a choice observer allows them to see when that choice is exercised.
```
Comment (bame-da) [inline docs/source/concepts/glossary.rst:116]:
```suggestion
A **choice observer** is a `party <#party>`__ on a `choice <#choice>`__. Choice observers are guaranteed to see the choice being exercised and all its consequences with it.
```

File: docs/source/daml/reference/structure.rst [hunk 8]
```diff
@@ -95,6 +95,15 @@ Here's the structure of a choice inside a template. There are two ways of specif
 
     Who can exercise the choice.
 
+:ref:`choice observers <daml-ref-choice-observers>`
+    ``observer`` keyword
+
+    Optional. Parties that aren't stakeholder but who you still want to be able to see this choice.
```
Comment (bame-da) [inline docs/source/daml/reference/structure.rst:101]:
```suggestion
    Optional. Additional parties that are guaranteed to be informed of an exercise of the choice.
```

**Review by bame-da — COMMENTED** [review]
If you already add docs to docs.daml.com, please annotate them as "Early Access" and say they are only available if you target LF 1.dev. We have not yet documented generic maps afaik.

File: rules_daml/daml.bzl [hunk 9]
```diff
@@ -358,6 +360,7 @@ $$DAMLC test --files {files}
                 )
                 for src in srcs
             ]),
+            target = "--target=" + target if (target) else "",
```
Comment (hurryabit) [inline rules_daml/daml.bzl:363]: Wow, I'm surprised we didn't already have this.

File: docs/source/daml/reference/choices.rst [hunk 10]
```diff
@@ -27,6 +27,17 @@ The main difference is that starting with ``choice`` means that you can pass in
 
 In contrast, if you start with ``controller``, the ``controller`` is automatically added as an observer when you compile your DAML files.
 
+.. _daml-ref-choice-observers:
+
+A secondary difference is that starting with ``choice`` allows choice observers to be attached to the choice. They will see all consequences of the action, along with other stakeholders.
```
Comment (hurryabit) [inline docs/source/daml/reference/choices.rst:32]:
```suggestion
A secondary difference is that starting with ``choice`` allows *choice observers* to be attached to the choice using the ``observer`` keyword. The choice observers are a list of parties that, in addition to the stakeholders, will see all consequences of the action.
```

File: docs/source/daml/code-snippets-dev/Structure.daml [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+  where
+    signatory exampleParty
+
+    -- start of choice observer snippet
+    -- choice observers may be specified if option 1 is used
+    choice NameOfChoiceWithObserver :
+          () -- replace () with the actual return type
+        with
+        party : Party -- parameters here
+      observer party -- optional specification of choice observers
```
Comment (hurryabit) [inline docs/source/daml/code-snippets-dev/Structure.daml:27]:
```suggestion
      observer party -- optional specification of choice observers (currently only available in DAML-LF 1.dev)
```

**Review by hurryabit — APPROVED** [review]
This looks very good to me. Thank you so much!

File: rules_daml/daml.bzl (hunk 9 above)
Comment (nickchapman-da) [reply rules_daml/daml.bzl:363]: me too

**Pull request merged**